STYLE.md

```diff
@@ -0,0 +1,432 @@
+```
+
+Type classes in `deriving` section should be always surrounded by
+parentheses.
+
+_WARNING_: try to avoid aggressive autoderiving. Deriving instances can
+slowdown compilation
+(stated here: http://www.stephendiehl.com/posts/production.html)
+
+> Deriving instances of Read/Show/Data/Generic for largely recursive ADTs can
+> sometimes lead to quadratic memory behavior when the nesting gets deep.
+
+### Function declaration
+
+All functions must have type signatures.
+
+Specialize function type signature for concrete types if you're using this function
+with only one type for each argument. Otherwise you should use more polymorphic
+version. Compiler can optimize specialized functions better
+and meaning of this function may be clearer.
+Use this rule unless you are the library creator and want your library to be abstract
+as possible.
+
+It is allowed to omit parentheses for only one type class constraint.
+
+If function type signature is very long then place type of each argument under
+its own line with respect to alignment.
+
+```haskell
+putValueInState
+  :: MonadIO m
+  => UserState
+  -> Maybe Int
+  -> AppConfig
+  -> (Int -> m ())
+  -> m ()
+```
+
+If the line with argument names is too big then put each argument on its own line
+and separate it somehow from body section.
+
+```haskell
+putValueInState
+  userState
+  mValue@(Just x)
+  appConfig
+  valueModificator
+  = do
+    <code goes here>
+```
+
+### List Declarations
+
+Align the elements in the list.  Example:
+
+```haskell
+exceptions =
+  [ InvalidStatusCode
+  , MissingContentHeader
+  , InternalServerError
+  ]
+```
+
+### If-then-else clauses
+
+Generally, guards and pattern matches should be preferred over _if-then-else_
+clauses, where possible.  Short cases should usually be put on a single line
+(when line length allows it).
+
+### Case expressions
+
+The alternatives in a case expression can be indented using either of
+the two following styles:
+
+```haskell
+foobar =
+  case something of
+    Just j  -> foo
+    Nothing -> bar
+```
+
+or as
+
+```haskell
+foobar = case something of
+           Just j  -> foo
+           Nothing -> bar
+```
+
+Align the `->` arrows when it helps readability.
+
+Dealing with laziness
+---------------------
+
+By default, use strict data types and lazy functions.
+
+### Data types
+
+Constructor fields should be strict, unless there's an explicit reason
+to make them lazy.  This helps to avoid many common pitfalls caused by too much
+laziness and reduces the number of brain cycles the programmer has to
+spend thinking about evaluation order.
+
+```haskell
+-- Good
+data Point = Point
+  { pointX :: !Double  -- ^ X coordinate
+  , pointY :: !Double  -- ^ Y coordinate
+  }
+```
+
+```haskell
+-- Bad
+data Point = Point
+  { pointX :: Double  -- ^ X coordinate
+  , pointY :: Double  -- ^ Y coordinate
+  }
+```
+
+TODO: Talk to luke
+Additionally, unpacking simple fields often improves performance and
+reduces memory usage:
```

Comment:
Unpacking strict fields can be a good idea but sometimes it's not (see links below). I tend to only look at manually `UNPACK`ing when I'm specifically trying to reduce memory usage for a particular data type. This might be worth the mention but I'm not sure it's worth enforcing as part of the style. 

https://wiki.haskell.org/Performance/Data_types#Unpacking_strict_fields
https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#unpack-pragma

Reply:
Maybe @dcoutts will want to weigh in on this as well.

Reply:
@intricate  Perhaps we can leave this out of the style document as well and review on a case by case basis.